crates/control/src/foot_bumper_filter.rs

```diff
@@ -0,0 +1,208 @@
+    right_in_use: bool,
+    left_detection_buffer: VecDeque<bool>,
+    right_detection_buffer: VecDeque<bool>,
+    left_count: i32,
+    right_count: i32,
+    last_left_time: Option<SystemTime>,
+    last_right_time: Option<SystemTime>,
+    left_pressed_last_cycle: bool,
+    right_pressed_last_cycle: bool,
+}
+
+#[context]
+pub struct CreationContext {
+    pub buffer_size: Parameter<usize, "foot_bumper_filter.buffer_size">,
+    pub obstacle_distance: Parameter<f32, "foot_bumper_filter.obstacle_distance">,
+    pub sensor_angle: Parameter<f32, "foot_bumper_filter.sensor_angle">,
+}
+
+#[context]
+pub struct CycleContext {
+    pub acceptance_duration: Parameter<Duration, "foot_bumper_filter.acceptance_duration">,
+    pub activations_needed: Parameter<i32, "foot_bumper_filter.activations_needed">,
+    pub enabled: Parameter<bool, "obstacle_filter.use_foot_bumper_measurements">,
+    pub number_of_detections_in_buffer_for_defective_declaration: Parameter<
+        usize,
+        "foot_bumper_filter.number_of_detections_in_buffer_for_defective_declaration",
+    >,
+    pub number_of_detections_in_buffer_to_reset_in_use:
+        Parameter<usize, "foot_bumper_filter.number_of_detections_in_buffer_to_reset_in_use">,
+    pub obstacle_distance: Parameter<f32, "foot_bumper_filter.obstacle_distance">,
+    pub sensor_angle: Parameter<f32, "foot_bumper_filter.sensor_angle">,
+
+    pub cycle_time: Input<CycleTime, "cycle_time">,
+    pub fall_state: Input<FallState, "fall_state">,
+    pub sensor_data: Input<SensorData, "sensor_data">,
+
+    pub foot_bumper_values: AdditionalOutput<FootBumperValues, "foot_bumper_values">,
+}
+
+#[context]
+#[derive(Default)]
+pub struct MainOutputs {
+    pub foot_bumper_obstacle: MainOutput<Vec<FootBumperObstacle>>,
+}
+
+impl FootBumperFilter {
+    pub fn new(context: CreationContext) -> Result<Self> {
+        let left_point = point![
+            context.sensor_angle.cos() * *context.obstacle_distance,
+            context.sensor_angle.sin() * *context.obstacle_distance
+        ];
+        let right_point = point![
+            context.sensor_angle.cos() * *context.obstacle_distance,
+            -context.sensor_angle.sin() * *context.obstacle_distance
+        ];
+        let middle_point = point![*context.obstacle_distance, 0.0];
+        Ok(Self {
+            left_point,
+            right_point,
+            middle_point,
+            left_in_use: true,
+            right_in_use: true,
+            left_detection_buffer: VecDeque::from(vec![false; *context.buffer_size]),
+            right_detection_buffer: VecDeque::from(vec![false; *context.buffer_size]),
+            ..Default::default()
+        })
+    }
+
+    pub fn cycle(&mut self, mut context: CycleContext) -> Result<MainOutputs> {
+        let fall_state = context.fall_state;
+
+        if !context.enabled {
+            return Ok(MainOutputs::default());
+        }
+
+        if context.sensor_data.touch_sensors.left_foot_left
```

Comment:
Maybe extract `context.sensor_data.touch_sensors` in a local variable `touch_sensors`

Reply:
Done